### PR TITLE
[TRAFODION-1863] PR340 caused PushdownV2 issues with column having non null default

### PR DESCRIPTION
**Pull request by eowhadi**
since non null default columns are sometime resolved at select time, we need to disable pushdown for these.

### CODE REVIEW

**Comment by eowhadi** [comment]
jenkins, retest

File: core/sql/optimizer/ValueDesc.cpp [hunk 1]
```diff
@@ -529,14 +529,13 @@ NABoolean ValueId::isAddedColumnWithNonNullDefault() const{
   default:
       break;
   }
-  if (nac && nac->isAddedColumn() && nac->getDefaultValue())
+  if (nac &&  nac->getDefaultValue() && strcmp(nac->getDefaultValue(),"NULL") != 0)
```
Comment (selvaganesang) [inline core/sql/optimizer/ValueDesc.cpp]: Will it be better to check for DefaultColumnClass == COM_NULL_DEFAULT instead.

Comment (eowhadi) [inline core/sql/optimizer/ValueDesc.cpp]: why is it better? I know what I do work. I don't know about the alternative. But I know how much a re run of regression cost :-)

**Comment by selvaganesang** [comment]
Agree. But, isn't good to the use the existing methods. Moreover, the COM_NULL_DEFAULT is used in many places already.

generator/GenExpGenerator.cpp:    case COM_NULL_DEFAULT:
optimizer/NATable.cpp:                             COM_NULL_DEFAULT  ,//defaultClass,
smdio/CmUtil.cpp:  {COM_NULL_DEFAULT, COM_NULL_DEFAULT_LIT},
sqlcomp/CmpDescribe.cpp:  else if (nac->getDefaultClass() == COM_NULL_DEFAULT)
sqlcomp/CmpSeabaseDDLcommon.cpp:          defaultClass = COM_NULL_DEFAULT;
sqlcomp/CmpSeabaseDDLcommon.cpp:              defaultClass = COM_NULL_DEFAULT;
sqlcomp/CmpSeabaseDDLtable.cpp:      colInfoArray[index].defaultClass = COM_NULL_DEFAULT;
sqlcomp/CmpSeabaseDDLtable.cpp:      else if (colInfo.defaultClass == COM_NULL_DEFAULT)
sqlcomp/CmpSeabaseDDLutil.cpp:  {COM_NULL_DEFAULT, COM_NULL_DEFAULT_LIT},

**Comment by eowhadi** [comment]
OK so tell me if this change is correct and is better:
replace: if (nac &&  nac->getDefaultValue() && strcmp(nac->getDefaultValue(),"NULL") != 0)

by: if (nac && nac-> getDefaultClass() != COM_NO_DEFAULT &&
nac->getDefaultClass() != COM_ALWAYS_COMPUTE_COMPUTED_COLUMN_DEFAULT &&
nac->getDefaultClass() != COM_ALWAYS_DEFAULT_COMPUTED_COLUMN_DEFAULT &&
nac->getDefaultClass() != COM_NULL_DEFAULT)

or probably using a switch construct with same semantic...

**Comment by selvaganesang** [comment]
I think the exact replacement of the existing code will be (nac && nac->getDefaultValue() && (nac->getDefaultClass() != COM_NULL_DEFAULT).

Returns TRUE for any column with the default value but with non-null default value. This is what I understood reading the code before. For anything else it would be false. 

You can add nac->getDefaultClass() != COM_NULL_DEFAULT && nac->getDefaultClass() != COM_CURRENT_DEFAULT() because the columns with current default will be in hbase table always even when the column is omitted.

File: core/sql/generator/GenPreCode.cpp [hunk 2]
```diff
@@ -11386,7 +11386,7 @@ NABoolean HbaseAccess::isHbaseFilterPredV2(Generator * generator, ItemExpr * ie,
   }
   //check if not an added column with default non null
   if ((foundBinary || foundUnary)&& (NOT hbaseLookupPred)){
-        if (colVID.isAddedColumnWithNonNullDefault()){
+        if (colVID.isColumnWithNonNullDefault()){
```
Comment (zellerh) [inline core/sql/generator/GenPreCode.cpp]: Would a check for added columns with non-null default be sufficient for aligned format?

File: core/sql/optimizer/ValueDesc.cpp (hunk 1 above)
Comment (zellerh) [inline core/sql/optimizer/ValueDesc.cpp]: Agree with Selva that checking the enum is safer and more maintainable. Maybe you can make this change in a future check-in, so you don't have to rerun the regressions just for that?

**Comment by zellerh** [comment]
Looks good to me, with some small comments.

File: core/sql/generator/GenPreCode.cpp (hunk 2 above)
Comment (eowhadi) [inline core/sql/generator/GenPreCode.cpp]: this code is totally bypassed for align format

**Comment by eowhadi** [comment]
Test passed , I guess good for merging?

**Pull request merged**